subsys/net/lib/lwm2m/lwm2m_rd_client.c

```diff
@@ -229,6 +229,8 @@ static void sm_handle_failure_state(enum sm_engine_state sm_state)
 		event = LWM2M_RD_CLIENT_EVENT_REG_UPDATE_FAILURE;
 	} else if (client.engine_state == ENGINE_DEREGISTER_SENT) {
 		event = LWM2M_RD_CLIENT_EVENT_DEREGISTER_FAILURE;
+	} else {
+		event = LWM2M_RD_CLIENT_EVENT_NETWORK_ERROR;
```

Comment:
Did you encounter any problems with the existing failure notification implmentation?

We have a simple backoff mechanism implemented (https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/lwm2m/lwm2m_rd_client.c#L162) to prevent from this event being notified on each socket failure.

The idea was to try to recover first (i. e. by retrying the registration) in such case, and notify the error to the application only after several failed attempts, so that the application can then act appropriately (for instance try to reconnect to the cellular network and restart the lwm2m engine). Current proposal will bypass this mechanism, and notify every single socket error. 

What's your say, do you think it'd be more appropriate to notify the application on every socket error?

Reply:
The reason (to give you some context) for the changes in this PR is actually tied a lot to existing draft PR [#41394](https://github.com/zephyrproject-rtos/zephyr/pull/41394). 
In that PR, I introduced client connection state for keeping track when is client online or offline. If you peek the PR, you'll notice I set client conn status to OFFLINE when there's either timeout or failure, which is important so observers don't get deleted by subsequent state machine restarts to DO_REGISTRATION state where context is cleared.

Now that being said, I see your point, but I think socket fault should be handled transparently and reported to application. Application can then opt-in for network errors or ignore them and handle only lwm2m operation errors (reg fails, update fails etc). AFAIS, this PR doesn't break silent retry mechanism? I was thinking about removing it as whole, but application can stop client at any point so I thought it was not necessary.

Reply:
Thank you for elaborating. No, I don't think it breaks the retry mechanism, I think though that we should refactor it a bit, in case we want to notify each error to application.

First of all, we should get rid of `CONFIG_LWM2M_RD_CLIENT_MAX_RETRIES` and event notification in case the counter is reached(https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/lib/lwm2m/lwm2m_rd_client.c#L162)

Next, I think it'd make sense to reset the `client.retries` counter after a successful connection, so in case we recover and encounter socket error again, we start counting fresh (it'd be needed to determine the backoff time properly).

Finally, some maximum backoff value should be specified (UINT8_MAX perhaps), so that if the application decides to ignore the event, we don't overflow the backoff counter (as it's uint8_t only).

That way, if the application decides to ignore socket errors, we will try to recover with some delay between consecutive connection attempts, until sucessfull.

Reply:
I agree. Also I've noticed that CONFIG_LWM2M_RD_CLIENT_MAX_RETRIES is not bounded by any interval, which means if it's configured with value bigger than 8, `client.retry_delay` would overflow since `1<<8` overflows UINT8_MAX.

One more problem with current implementation is that `client.retry_delay` holds counter which is decreased on each run of `lwm2m_rd_client_service` in `ENGINE_NETWORK_ERROR` state which uses hardcoded value 500 milliseconds. 
So this means maximum backoff delay is `1<<7 / 2 = 64` seconds. 

In our case, on top of this built-in retry mechanism, we are using custom backoff mechanism because we need much higher delay values in minute, hours or even day long range. This is especially important in low power networks like NBIoT where power consumption is priority. 
I would also watch out for not introducing unnecessary complexity to engine, as it is we have two retry mechanisms:

- 4 retries on COAP layer in case of timeout (4 is recommended as default value by RFC)
- CONFIG_LWM2M_RD_CLIENT_MAX_RETRIES in case of network error. Note that client only reaches this state when it fails to send registration (or bootstrap) because that's where sockets are created/started. 

Another problem with current state of `socket_fault_cb` is that we are jumping directly to registration phase in case of error. If client uses a long lifetime interval, e.g. 24h, and socket fault happens on notification after 4h, because device lost signal, there's high possibility network will recover and registration update will succeed (I'm ignoring IP address change here but that's another problem). Our current solution is to have client entered OFFLINE state and in that case notifications are stored directly to storage medium without even attempting to send them, and we would wait for registration update to resend them. Or we could trigger registration update when network recovers.

One more important aspect of handling client is that we can't stop it at any point because that stops engine thread which takes care of notifications. But this is specific to our use case, I wanted to keep this PR relevant to general usage of the lwm2m.

Reply:
I see, thank you for highlighting your usage of the library, this gives me way better understanding of the problem.

If I understand correctly, it seems that this internal recovery mechanism in case of network errors might not be desired at all. I think we should give more responsibility to the application, i. e. only forward the fault event w/o taking any action in the library. The application could then decide, whether to restart the engine or take any other action. This also means any backoff mechanisms should be implemented on the application side to match its needs.

We should also get rid of `ENGINE_NETWORK_ERROR` state in this case, as there'd be no longer use for this state.

Do you agree with the above, or did I mix up things? @SeppoTakalo Feel free to chip in if you have an opinion on this.

Reply:
> If I understand correctly, it seems that this internal recovery mechanism in case of network errors might not be desired at all.

Yes, that was my point exactly.

> We should also get rid of `ENGINE_NETWORK_ERROR` state in this case, as there'd be no longer use for this state.

Not necessarily, I think we should keep network error, maybe rename it to SOCKET_ERROR, because engine is not aware of network status, only socket status.

Another point I would like to discuss is socket usage. I'm not familiar with it all that much, but I assume when there is a socket error there is no recovery in place right now? To continue on my previous use case regarding notification fail and keeping the engine running, what would happen once network reconnects and we would like to trigger registration update? Do the sockets need to be restarted/recreated in order for communication to work seamlessly?

Reply:
To me, it sounds like we need to rethink the engine recovery and back-off logic a bit more.
I'm not actually totally sure how it works now, so I'm not ready to say how this PR should be made.

But I would like to point out that few corner cases need to be considered:
* Non secure UDP.
* DTLS

Both behave differently. This is because on UDP, when there is any socket error, we know that it is local. There is no messages coming back from lost UDP packets. There is no reason to retry immediately. We should back off by some time to wait for local connection to heal. Any dropped UDP packets should be handled by CoAP layer. So it should retry, if needed. Actually, thinking of this, there is probably no reason to add yet another retrying layer after the CoAP, so we could just back off similarly, if CoAP layer reports timeout.

For DTLS, there is retry and timeout logic on handshake phase. At one point it looked like it retries six times before returning us that timeout have happened. Does it make sense for our engine to retry again? Probably not.
Once the connection is established, it behaves like UDP, dropped packets should be handled by CoAP layer.
But in DTLS case, I don't know how it translates the socket errors. It probably passes those through to engine.

After all, I'm starting to feel like on every timeout or socket error, we should just drop the socket. Then later on recover the connection with the new socket, just so that DTLS sockets get properly initialized.
But we should still maintain the registration time and status. If we happen to recover the connection we should just send the registration update, and so keep all the subscriptions active. It should be OK even if the IP+port changes. I think this is intentional for queue mode.

What comes to notify-messages. I'm not in favor of storing them. I would drop, or actually just delay the send until we are connected. If we follow the data model of LwM2M, then latest data is stored in the resources, and we should always send the latest data.

Reply:
> > Not necessarily, I think we should keep network error, maybe rename it to SOCKET_ERROR, because engine is not aware of network status, only socket status.
> 
> What would be the transition condition out of the error state then?

I see I didn't think this through fully. You're correct, ENGINE_NETWORK_ERROR state should be removed. SOCKET_ERROR would be a new event, not state, which is dispatched in case of socket error.

Reply:
> One more important aspect of handling client is that we can't stop it at any point because that stops engine thread which takes care of notifications. But this is specific to our use case, I wanted to keep this PR relevant to general usage of the lwm2m.

This is actually false, stopping client doesn't stop the thread. Stopping the client just causes the thread to enter IDLE mode. If deregister flag is set, client will perform deregistration and if successful, it will clear the context which deletes underlying socket. When there is no active socket (`sock_nfds < 1`), engine will basically wait for new socket to be created. 
Just wanted to write this here so there's no confusion. This has impact on historical notifications implementation since when there's no valid socket, `check_notifications` is never called and notifications are never stored.